### PR TITLE
Add: Heatmap chart visualization by Plotly

### PR DESCRIPTION
**Pull request by deecay**
redash_heatmap_in_dashboard

Editor:
redash_heatmap_editor

### CODE REVIEW

**Comment by deecay** [comment]
You are right. Should be fixed.

**Comment by kocsmy** [comment]
@deecay can you make sure this fits to the newly introduced color scheme [here](https://github.com/getredash/redash/pull/2282) (and you're starting from the default red and blue and altering them respectively). 

I'm happy to help here and give you specific hex codes that matches if you need assistance.

**Comment by deecay** [comment]
@kocsmy do you have an idea of what kind of gradient should the default be?
Blue: '#356AFF' to  Red: '#E92828'?
Or something like LightGreen - Green - Green2 - Green3?

**Comment by kocsmy** [comment]
hey @deecay — I'd introduce new shades starting from the base colors:
screenshot 2018-03-02 12 31 51

**Comment by kocsmy** [comment]
If you can let me know how many colors do we need exactly and what should be the base colors I can prepare them for you!

**Comment by kravets-levko** [comment]
How about adding two options - allow user to select two colors, and then  use them for `colorscale` (`[[0, 'color-from', [1, 'color-to`)? WDYT?

https://plot.ly/javascript/reference/#heatmap-colorscale

**Comment by kravets-levko** [comment]
BTW - it's possible to set more than two colors, like:
`[ [0, color1], [0.3, color2], [0.7, color3], [1, color4] ]`

**Comment by kocsmy** [comment]
@kravets-levko I like that idea, sounds good to me.

**Comment by deecay** [comment]
Just sharing how it looks.

* 2-color: `[[0, '#356aff'], [1, '#e92827']]`
image

* 8-color: `[[0, '#356aff'], [0.14, '#4a7aff'], [0.28, '#5d87ff'], [0.42, '#7398ff'], [0.56, '#fb8c8c'], [0.71, '#ec6463'], [0.86, '#ec4949'], [1, '#e92827']]`
image

I will make either of them the default, and then add interface to specify `from-color` and `to-color`. Any inputs are welcomed.

**Comment by arikfr** [comment]
I think that the 8-color option is a better default.

**Comment by kravets-levko** [comment]
If it makes sense, both variants may be implemented: dropdown with list of pre-defined themes, and with item "Custom". When selecting "Custom" user can select two colors.

**Comment by deecay** [comment]
@arikfr ok. done.
@kravets-levko this is done too.

Color selection are done by text inputs to allow any color. Do we need color selector as in series color, or any other color-picker?

**Comment by kravets-levko** [comment]
@deecay IMHO user shouldn't be able to select arbitrary colors; we have predefined color list that we use for other plotly-based charts ([here](https://github.com/getredash/redash/blob/master/client/app/visualizations/chart/plotly/utils.js#L9-L34)) and we already use color selector [here](https://github.com/getredash/redash/blob/master/client/app/visualizations/chart/chart-editor.html#L242-L249). So, I think, we should keep using it in other places too. @arikfr what do you think?

**Comment by deecay** [comment]
@kravets-levko This is how the editor looks with color selector.

heatmap color picker

**Comment by kravets-levko** [comment]
@deecay Looks incredible! :-) I think both color selectors could be placed on a single line:
```html
<div class="row">
   <div class="col-xs-6"><!-- color picker 1 --></div>
   <div class="col-xs-6"><!-- color picker 2 --></div>
</div>
```

**Comment by deecay** [comment]
now it looks like this.

image

File: client/app/visualizations/chart/chart-editor.html [hunk 1]
```diff
@@ -62,7 +62,7 @@
 
     </div>
 
-    <div class="form-group" ng-if="options.globalSeriesType != 'custom'">
+    <div class="form-group" ng-if="options.globalSeriesType != 'custom' && options.globalSeriesType != 'heatmap'">
```
Comment (kravets-levko) [inline client/app/visualizations/chart/chart-editor.html]: Replace it with `"['custom', 'heatmap'].indexOf(options.globalSeriesType) >= 0"` - it's easier to read and modify in future

File: client/app/visualizations/chart/chart-editor.html [hunk 2]
```diff
@@ -99,7 +111,7 @@
       </ui-select>
     </div>
 
-    <div class="checkbox" ng-if="options.globalSeriesType != 'custom'">
+    <div class="checkbox" ng-if="options.globalSeriesType != 'custom' && options.globalSeriesType != 'heatmap'">
```
Comment (kravets-levko) [inline client/app/visualizations/chart/chart-editor.html]: Replace it with `"['custom', 'heatmap'].indexOf(options.globalSeriesType) >= 0"` - it's easier to read and modify in future

File: client/app/visualizations/chart/chart-editor.html [hunk 3]
```diff
@@ -113,7 +125,7 @@
       </label>
     </div>
 
-    <div class="form-group" ng-if="options.globalSeriesType != 'custom'">
+    <div class="form-group" ng-if="options.globalSeriesType != 'custom' && options.globalSeriesType != 'heatmap'">
```
Comment (kravets-levko) [inline client/app/visualizations/chart/chart-editor.html]: Replace it with `"['custom', 'heatmap'].indexOf(options.globalSeriesType) >= 0"` - it's easier to read and modify in future

File: client/app/visualizations/chart/chart-editor.html [hunk 4]
```diff
@@ -133,6 +145,47 @@
         </label>
       </div>
     </div>
+    <div class="form-group" ng-if="options.globalSeriesType == 'heatmap'">
+      <label class="control-label">Color Scheme</label>
+
+      <ui-select ng-model="options.colorScheme"">
+        <ui-select-match allow-clear="true" placeholder="Choose Colorscheme...">{{$select.selected | capitalize}}</ui-select-match>
```
Comment (kravets-levko) [inline client/app/visualizations/chart/chart-editor.html]: "Colorscheme" -> "Color Scheme"

File: client/app/visualizations/chart/index.js [hunk 5]
```diff
@@ -87,7 +88,12 @@ function ChartEditor(ColorPalette, clientConfig) {
       };
       scope.chartTypeChanged();
 
+      scope.colorScheme = ['Blackbody', 'Bluered', 'Blues', 'Earth', 'Electric',
+        'Greens', 'Greys', 'Hot', 'Jet', 'Picnic', 'Portland',
+        'Rainbow', 'RdBu', 'Reds', 'Viridis', 'YlGnBu', 'YlOrRd', 'Custom..'];
```
Comment (kravets-levko) [inline client/app/visualizations/chart/index.js]: `'Custom..'` -> `'Custom...'` (thee dots) + fix value in `ng-if`s in editor template and js code

File: client/app/visualizations/chart/chart-editor.html (hunk 3 above)
Comment (kravets-levko) [reply client/app/visualizations/chart/chart-editor.html]: Sorry, my bad. Yes, `== -1`

**Comment by kocsmy** [comment]
Very nice job on this one!

**Comment by deecay** [comment]
Hi,

Could you look into this PR?

I have moved color related controls from General Tab to Colors tab, and also made Number Values Format from Data Labels effective.

**Comment by arikfr** [comment]
@kravets-levko is this good to merge?

**Comment by kravets-levko** [comment]
@arikfr Yes, and I think we finally have to merge this! @deecay - it's incredible, very, very nice job!

**Comment by arikfr** [comment]
Thanks, @deecay . Sorry for the long wait.

**Pull request merged**